contracts/marketpools.js

```diff
@@ -0,0 +1,348 @@
+    quoteQuantity,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    || !api.assert(baseQuantity && api.BigNumber(baseQuantity).gt(0), 'invalid baseQuantity')
+    || !api.assert(quoteQuantity && api.BigNumber(quoteQuantity).gt(0), 'invalid quoteQuantity')
+    || !await validateTokenPair(tokenPair)) {
+    return;
+  }
+
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  if (api.assert(pool, 'no existing pool for tokenPair')) {
+    if (!api.assert(api.BigNumber(pool.baseQuantity).gt(0)
+      && api.BigNumber(pool.quoteQuantity).gt(0), 'insufficient liquidity')
+      || !validateLiquiditySwap(pool, baseQuantity, quoteQuantity)) return;
+
+    const lp = await api.db.findOne('liquidityPosition', { account: api.sender, tokenPair });
+    if (api.assert(lp, 'no existing liquidity position')
+      && api.assert(api.BigNumber(lp.baseQuantity).minus(baseQuantity).gte(0)
+      && api.BigNumber(lp.quoteQuantity).minus(quoteQuantity).gte(0), 'not enough liquidity in position')) {
+      lp.baseQuantity = api.BigNumber(lp.baseQuantity).minus(baseQuantity);
+      lp.quoteQuantity = api.BigNumber(lp.quoteQuantity).minus(quoteQuantity);
+      if (api.BigNumber(lp.baseQuantity).eq(0) && api.BigNumber(lp.quoteQuantity).eq(0)) {
+        await api.db.remove('liquidityPosition', lp);
+      } else {
+        await api.db.update('liquidityPosition', lp);
+      }
+
+      await api.transferTokens(api.sender, baseSymbol, baseQuantity, 'user');
+      await api.transferTokens(api.sender, quoteSymbol, quoteQuantity, 'user');
+      updatePoolStats(pool, -baseQuantity, -quoteQuantity, false);
+      api.emit('removeLiquidity', { memo: `Remove ${baseSymbol} and ${quoteSymbol}` });
+    }
+  }
+};
+
+actions.swapTokensForExactTokens = async (payload) => {
+  const {
+    tokenPair,
+    tokenSymbol,
+    tokenOut,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    || !api.assert(typeof (tokenSymbol) === 'string', 'invalid token')
+    || !api.assert(tokenOut && api.BigNumber(tokenOut).gt(0), 'insufficient tokenOut')
+    || !await validateTokenPair(tokenPair)) {
+    return;
+  }
+
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  let liquidityIn;
+  let liquidityOut;
+  let symbolIn;
+  let symbolOut;
+  if (tokenSymbol !== baseSymbol) {
+    liquidityIn = pool.baseQuantity;
+    liquidityOut = pool.quoteQuantity;
+    symbolIn = baseSymbol;
+    symbolOut = quoteSymbol;
+  } else {
+    liquidityIn = pool.quoteQuantity;
+    liquidityOut = pool.baseQuantity;
+    symbolIn = quoteSymbol;
+    symbolOut = baseSymbol;
+  }
+  if (!api.assert(pool, 'no existing pool for tokenPair')) return;
```

Comment:
this assertion should be moved up to right after the findOne statement.

Reply:
Fixed